File: src/main/java/player/ki/StaticSpaceEvaluation.java
package player.ki;

import board.ChessGameInterface;
import figures.Figure;
import helper.Position;

import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

/**
 * @author stephan
 */
class StaticSpaceEvaluation implements StaticEvaluationInterface {

    public Evaluaded evaluate(ChessGameInterface game, boolean forWhite) {
        final int BASIC_VALUE = 200;
        Position whiteKingPos = null;
        Position blackKingPos = null;
        boolean colorOfWinnerIsWhite = true;
        List figuresPos = new LinkedList();

        for (int row = 0; row < 8; row++) {
            for (int column = 0; column < 8; column++) {
                Position pos = Position.get(row, column);
                if (!game.isFreeArea(pos)) {
                    Figure figure = game.getFigure(pos);
                    if (figure.isKing()) {
                        if (figure.isWhite()) whiteKingPos = pos;
                        else blackKingPos = pos;
                    } else {
                        colorOfWinnerIsWhite = figure.isWhite();
                        figuresPos.add(pos);
                    }
                }
            }
        }

        float value = 0;
        Position kingPos;
        if (colorOfWinnerIsWhite) {
            figuresPos.add(whiteKingPos);
            kingPos = blackKingPos;
        } else {
            figuresPos.add(blackKingPos);
            kingPos = whiteKingPos;
        }

        value = countRestSpace(game, kingPos, figuresPos);
        value += 3 * borderMalus(kingPos);
        value += 5 * getKingDistance(whiteKingPos, blackKingPos);
        value -= BASIC_VALUE;

        if (colorOfWinnerIsWhite == forWhite) value = -value;

        //TODO is there is better secondary evaluation that a static 0?
        return Evaluaded.fromValues(value / 10.0f, 0);
    }

    int getKingDistance(Position pos1, Position pos2) {
        //return	Math.max( Math.abs(pos1.row-pos2.row),Math.abs(pos1.column-pos2.column) );
        return Math.abs(pos1.row - pos2.row) + Math.abs(pos1.column - pos2.column);
    }

    float borderMalus(Position kingPos) {
        return -Math.max(Math.abs(kingPos.row - 3.5f), Math.abs(kingPos.column - 3.5f));
    }

    int countRestSpace(ChessGameInterface game,
                       Position kingPos,
                       List otherFiguresPos) {
        boolean figuresHaveWhiteColor = !game.getFigure(kingPos).isWhite();
        HashSet foundPositions = new HashSet(64);
        HashSet searchPositions = new HashSet(2);
        foundPositions.add(kingPos);
        searchPositions.add(kingPos);
        while (!searchPositions.isEmpty()) {
            HashSet newSearchPositions = new HashSet(10, 1);
            Iterator iter = searchPositions.iterator();
            while (iter.hasNext()) {
                Position search_this = (Position) iter.next();
                Iterator newIter = getPossibleKingPositions(search_this);
                while (newIter.hasNext()) {
                    Position foundPos = (Position) newIter.next();
                    if (!foundPositions.contains(foundPos) &&
                            isNotReachableByFiguresInList(foundPos,
                                    otherFiguresPos,
                                    game,
                                    figuresHaveWhiteColor)
                            ) {
                        foundPositions.add(foundPos);
                        newSearchPositions.add(foundPos);
                    }
                }
            }
            searchPositions = newSearchPositions;
        }

        return foundPositions.size();
    }

    private boolean isNotReachableByFiguresInList(Position to,
                                                  List other,
                                                  ChessGameInterface game,
                                                  boolean figuresHaveWhiteColor) {
        Iterator iter = other.iterator();
        while (iter.hasNext()) {
            Position from = (Position) iter.next();
            if (from.equalsPosition(to)) return false;
            if (game.isMoveable(from, to, figuresHaveWhiteColor)) return false;
        }
        return true;
    }

    private Iterator getPossibleKingPositions(Position pos) {
        int minRow = Math.max(0, pos.row - 1);
        int maxRow = Math.min(7, pos.row + 1);
        int minColumn = Math.max(0, pos.column - 1);
        int maxColumn = Math.min(7, pos.column + 1);

        LinkedList positions = new LinkedList();
        for (int row = minRow; row <= maxRow; row++) {
            for (int column = minColumn; column <= maxColumn; column++) {
                Position newPos = Position.get(row, column);
                if (!newPos.equalsPosition(pos)) {
                    positions.add(Position.get(row, column));
                }
            }
        }

        return positions.iterator();
    }

    static boolean shouldUseStaticSpaceEvaluation(ChessGameInterface game) {
        int whiteFigures = 0;
        int blackFigures = 0;
        for (int column = 0; column < 8; column++) {
            for (int row = 0; row < 8; row++) {
                Position pos = Position.get(row, column);
                if (!game.isFreeArea(pos)) {
                    Figure figure = game.getFigure(pos);
                    if (figure.isWhite()) {
                        whiteFigures++;
                    } else {
                        blackFigures++;
                    }
                    if (figure.isPawn() || (whiteFigures > 1 && blackFigures > 1)) return false;
                }
            }
        }
        return true;
    }
}

File: src/main/kotlin/voidchess/player/ki/evaluation/StaticSpaceEvaluation.kt
package voidchess.player.ki.evaluation

import voidchess.board.ChessGameInterface
import voidchess.board.forAllFigures
import voidchess.figures.Figure
import voidchess.board.move.Direction
import voidchess.board.move.Position

import java.util.HashSet
import java.util.LinkedList


class StaticSpaceEvaluation : StaticEvaluationInterface {

    override fun getPrimaryEvaluation(game: ChessGameInterface, forWhite: Boolean): Ongoing {
        val basicValue = 200
        val whiteKing = game.whiteKing
        val blackKing = game.blackKing

        var colorOfWinnerIsWhite = true
        val figuresPos = LinkedList<Position>()

        game.forAllFigures { figure ->
            if (!figure.isKing()) {
                colorOfWinnerIsWhite = figure.isWhite
                figuresPos.add(figure.position)
            }
        }

        val king = if (colorOfWinnerIsWhite) blackKing else whiteKing
        if (colorOfWinnerIsWhite) {
            figuresPos.add(whiteKing.position)
        } else {
            figuresPos.add(blackKing.position)
        }

        var value = countRestSpace(game, king, figuresPos).toDouble()
        value += 3 * borderMalus(king.position)
        value += (5 * whiteKing.position.distanceTo(blackKing.position)).toDouble()
        value -= basicValue.toDouble()

        if (colorOfWinnerIsWhite == forWhite) value = -value

        //TODO is there is better secondary evaluation that a static 0?
        return Ongoing(value / 10.0)
    }

    override fun addSecondaryEvaluation(game: ChessGameInterface, forWhite: Boolean, evaluated: Evaluated) {
        //do nothing (keep the secondary value = 0)
        //TODO is there a good secondary matrix of success for space?
    }

    private fun borderMalus(kingPos: Position): Double {
        return -Math.max(Math.abs(kingPos.row - 3.5), Math.abs(kingPos.column - 3.5))
    }

    fun countRestSpace(game: ChessGameInterface,
                       king: Figure,
                       otherFiguresPos: List<Position>): Int {
        val figuresHaveWhiteColor = !king.isWhite
        val foundPositions = HashSet<Position>(64)
        var searchPositions = HashSet<Position>(2)
        foundPositions.add(king.position)
        searchPositions.add(king.position)
        while (searchPositions.isNotEmpty()) {
            val newSearchPositions = HashSet<Position>(10, 1f)
            for (searchPos in searchPositions) {
                val newIter = getPossibleKingPositions(searchPos)
                while (newIter.hasNext()) {
                    val foundPos = newIter.next()
                    if (!foundPositions.contains(foundPos) && isNotReachableByFiguresInList(
                                    foundPos,
                                    otherFiguresPos,
                                    game,
                                    figuresHaveWhiteColor)) {
                        foundPositions.add(foundPos)
                        newSearchPositions.add(foundPos)
                    }
                }
            }
            searchPositions = newSearchPositions
        }

        return foundPositions.size
    }

    private fun isNotReachableByFiguresInList(to: Position,
                                              other: List<Position>,
                                              game: ChessGameInterface,
                                              figuresHaveWhiteColor: Boolean): Boolean {
        for (anOther in other) {
            if (anOther.equalsPosition(to)) return false
            if (game.isMovable(anOther, to, figuresHaveWhiteColor)) return false
        }
        return true
    }

    private fun getPossibleKingPositions(pos: Position): Iterator<Position> {
        val positions = LinkedList<Position>()
        for(direction in Direction.values()) {
            pos.step(direction)?.let {newPos ->
                positions.add(newPos)
            }
        }
        return positions.iterator()
    }

    companion object {

        fun shouldUseStaticSpaceEvaluation(game: ChessGameInterface): Boolean {
            var whiteFigures = 0
            var blackFigures = 0
            game.forAllFigures { figure ->
                if (figure.isWhite) {
                    whiteFigures++
                } else {
                    blackFigures++
                }
                if (figure.isPawn() || whiteFigures > 1 && blackFigures > 1) return false
            }
            return true
        }
    }
}
